config.yaml

```diff
@@ -1,14 +1,42 @@
-# This file defines charm config options, and populates the Configure tab on Charmhub.
-# If your charm does not require configuration options, delete this file entirely.
-#
-# See https://juju.is/docs/config for guidance.
-
 options:
-  role:
-    description: Role for this microk8s deployment
-    default: ""
+  addons:
+    description: |
```

Comment:
This should be refactored imo -- perhaps an option per addon? It would make config for each addon more obvious. 

```bash
juju config microk8s \
    addon_metallb=true \
    addon_metallb_range="10.42.23.10-10.42.23.20" 
    addon_dns=true
```

Thoughts?

Reply:
This was considered initially as well and is a great idea, especially for the core addons, as it helps put most things on rails. However, the implementation could then suffer:

- we have to basically iterate over all addons manually to manage them
- we have to keep the charm up to date when a new microk8s addon is added, or when new addon configuration options are added
- how do we deal with community addons with this approach?
- how do we deal with custom addon repositories in this case?

I also feel like most of this should be built on relations instead of enumerating configuration options. And this is also the end goal, to eventually be able to re-use the existing (and any future) charms like [coredns](https://charmhub.io/coredns), [metallb](https://charmhub.io/metallb-controller), [cilium](https://charmhub.io/cilium), etc) for more advanced use cases and configuration instead of having 2+ competing implementations with addons and hard-coded configuration options.

Ideas @jnsgruk @AlexsJones?

Reply:
We aren't ready to solution addons right here, it is out of the scope of this deliverable - I want to make sure we do a proper job and that's at the product level too, to ensure it's a consistent and coordinated message.

 I would advise we drop charmed microk8s from having addon configuration until it's been properly discussed and specced out. I will make sure to discuss to get Cedric's take on this too from a commercial/end-user POV before we start implementing

Reply:
> we have to keep the charm up to date when a new microk8s addon is added, or when new addon configuration options are added

Surely this is implicit? While I understand it's a little more work, in reality each time a core addon is added, it should be *deliberately* added to the charm and tested?

> how do we deal with community addons with this approach?
> how do we deal with custom addon repositories in this case?

These should both be handled through relations. We should probably have some sort of `microk8s-addon-repo` charm that takes some config for a repo, relates to `microk8s` and passes the details over the relation.

> Thus, it feels like having a single well-documented addons option is a good way forward for now.

This I disagree with. Once people start using this option, it'll be very difficult to move away from. Design the system you want to end up with and then let's implement that? :)

Reply:
I should add - it's fine for you to go incrementally here. If the first "new" version of the charm *only* supports core addons, then I think that's fine.

Designing the flow for community addons and custom repos should probably be specced and reviewed seperately.

Reply:
> This I disagree with. Once people start using this option, it'll be very difficult to move away from. Design the system you want to end up with and then let's implement that? :)

+1000. To add to that, if a single option (`addons`) is not the right way and it's dangerous for people to start using and depending on (which, I have to say, I more than stand behind as a decision), wouldn't tens of options (`addon_dns`, ..., `addon_metallb_range`, ..., `addon_gpu_set_as_default_runtime`, ...) fall into the same problem, but worse?

Frankly, this is the only reason why we're _keeping_ `addons` for now, so that people familiar with the `microk8s enable` flow be able to adjust.

Reply:
Well, *sort of*. 

The problem with freetext options like this one is that validation is tricky, and it's pretty easy for the charm to get into a borked state, and it's quite difficult to surface the right information to the user about how to un-bork it.

The point of the operator is to make operations seamless and "bulletproof". If we include config for the charm that allow the user to accidentally bork the charm, we've sort of failed at that.

Reply:
FWIW I prefer the explicit config options to enable and configure specific addons - for a-lot of the reasons already described above. 

We (Sunbeam) do use the addons option in the current microk8s charm (hostpath-storage, dns, metallb if that would help with priorities) but splitting these out into individual config options is trivial and much preferred.

Reply:
OK, from this conversation, my read is that _everything_ addon related should be kept outside of the main `microk8s` charm and into something separate, communicated to the charm over a `microk8s-addon` relation. We could then have subordinate charms:

- `microk8s-core-addons`, with toggle options our core addons, e.g. `dns`, `rbac`, `ingress`, `gpu`, `gpu_version`, etc. We could have individual tracks for different microk8s versions, and we avoid cluttering the `microk8s` charm with any of this.

- `microk8s-custom-addons`, which would allow user-defined addons and addon repositories.

This grants us agility, prevents us from adding cruft to the main charm from the get-go, and is compatible with being able to use charms for advanced use-cases instead (simply deploy the charm instead of enabling the addon). If we all agree that it goes into the right direction, we can move this discussion to a spec document and finalize the details there.

Reply:
Yeh I think the subordinate approach is a reasonable one. Give me a shout when you've got a spec together - it should probably cover versioning/compatibility and what the config schema will be. 

Imagine that @javacruft would like a look too :)

Reply:
Removed the addon handling from this PR, will revisit separately.